firedrake/output.py

```diff
@@ -464,6 +470,37 @@ def _prepare_output(self, function, max_elem):
 
         return OFunction(array=get_array(output), name=name, function=output)
 
+    def _prepare_coordinates(self, mesh, max_elem):
+        from firedrake import Function, Interpolator
+        from tsfc.finatinterface import create_element as create_finat_element
+
+        function = mesh.coordinates
+        name = function.name()
+        # Need to project/interpolate?
+        # If space is not the max element, we must do so.
+        finat_elem = function.function_space().finat_element
+        if finat_elem == create_finat_element(max_elem):
+            return OFunction(array=get_array(function),
+                             name=name, function=function)
+        #  OK, let's go and do it.
+        output = self._output_functions.get(function)
+        if output is None:
+            V = self._get_output_functionspace(function, max_elem)
+            output, interpolator = mesh.aux_coordinate_functions.get(V, (None, None))
+            if output is None:
+                # allocate a new coordinate function
+                print(f'Allocating coords for V={V}')
+                output = Function(V)
+                interpolator = Interpolator(function, output)
+                # store weak reference in the mesh
+                # need to store the interpolator object too
+                mesh.aux_coordinate_functions[V] = (weakref.proxy(output), weakref.proxy(interpolator))
+                interpolator.interpolate()
+            self._output_functions[function] = output
+            # FIXME interpolator is not used in File but we need to keep a strong reference in the File object?
+            self._mappers[function] = interpolator
```

Comment:
Presumably we need to store a reference of interpolator in File as well?

Reply:
yes.